### PR TITLE
Add support for iterative attestation by returning the end of the IMA measurement log 

### PR DESCRIPTION
**Pull request by stefanberger**
This PR adds support for iterative attestation by returning the end of the IMA measurement log starting at a certain point requested by the validator. This code is a translation of the code the python agent to Rust.

### CODE REVIEW

**Comment by ueno** [comment]
@stefanberger I tried both master (a58f5664e6062f514b57a12376b794c7775d19ae) and #269, but quote works without error:
```console
2021-10-21 05:18:03.340 - keylime.tenant - INFO - Quote from 192.168.127.2 validated
```
Additional info:
-  I've added a [slight modification for testing](https://github.com/keylime/rust-keylime/files/7386192/rust-keylime-testing.txt)
- verifier and registrar are from the keylime master (keylime/keylime@13eae5774e827084d3bd9ac518af23064d94a854)

**Comment by stefanberger** [comment]
> @stefanberger I tried both master ([a58f566](https://github.com/keylime/rust-keylime/commit/a58f5664e6062f514b57a12376b794c7775d19ae)) and #269, but quote works without error:

I am sure the agent would scream if one of the TPM commands was failing because of missing environment variable or so. I am not passing any TSS 2 related environment variables ... I thought maybe it could have something to do with issue #234 but this initial quote I think doesn't quote that many PCRs.

**Comment by ueno** [comment]
@stefanberger I switched to hardware TPM without TCTI envvars (i.e., `/dev/tpmrm0` is used), but it still works. How reproducible is it?

File: src/ima.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        if nth_entry >= self.tuples[1].0 {
+            return Some(self.tuples[1]);
+        }
+        return None;
+    }
+}
+
+// Singleton ImeMeasurementList
+static mut IMAMEASUREMENTLIST: ImaMeasurementList = ImaMeasurementList {
```
Comment (ueno) [inline src/ima.rs:47]: What about moving this to [`QuoteData`](https://github.com/keylime/rust-keylime/blob/master/src/main.rs#L89) struct in main.rs, which is kept during the agent lifetime?

File: src/quotes_handler.rs [hunk 2]
```diff
@@ -237,18 +248,32 @@ pub async fn integrity(
                 .await;
         };
 
+        let ima_ml_entry = req.uri().query().unwrap();
+        let nth_entry = match ima_ml_entry.find("ima_ml_entry=") {
+            None => 0,
+            Some(idx) => {
+                let number = &ima_ml_entry[idx + 13..];
+                let last: usize =
+                    number.find(|c: char| !c.is_numeric()).unwrap_or(0);
```
Comment (ueno) [inline src/quotes_handler.rs:257]: Shouldn't this be `.unwrap_or(number.len())`, if there are only numeric characters?

**Comment by stefanberger** [comment]
> @stefanberger I switched to hardware TPM without TCTI envvars (i.e., `/dev/tpmrm0` is used), but it still works. How reproducible is it?

This was my mistake and I had to clear the database due to a change in the EK cert. After clearing the databases I get to this error here:

```
# keylime_tenant -c add -t 192.168.122.16 -v 127.0.0.1 -u D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -f foo --exclude fcos/ExcludeList.appraiseonly --allowlist ./allowlist.json
Using config file /etc/keylime.conf
2021-10-21 12:27:39.995 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.3.0
2021-10-21 12:27:40.004 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2021-10-21 12:27:40.004 - keylime.registrar_client - WARNING - TLS is enabled.
2021-10-21 12:27:40.005 - keylime.registrar_client - INFO - Setting up client TLS...
/usr/lib/python3.9/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
2021-10-21 12:27:40.060 - keylime.registrar_client - CRITICAL - Error: unexpected http response code from Registrar Server: 404
2021-10-21 12:27:40.061 - keylime.registrar_client - CRITICAL - Response code 404: agent_id not found
2021-10-21 12:27:40.063 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x408000
2021-10-21 12:27:40.063 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x808000
2021-10-21 12:27:40.064 - keylime.ima - DEBUG - Loading allowlist from ./allowlist.json
2021-10-21 12:27:40.064 - keylime.ima - DEBUG - Loaded allowlist from ./allowlist.json with checksum 64ff0f11168fb684d3e9ef2ac486e03b6f6f2d82faf0590918a9260dd6905695
2021-10-21 12:27:40.064 - keylime.ima - DEBUG - Reading allow list as JSON format
2021-10-21 12:27:40.064 - keylime.ima - DEBUG - Allowlist has compatible version 3
2021-10-21 12:27:40.065 - keylime.ima - DEBUG - Loaded exclusion list from fcos/ExcludeList.appraiseonly: ['boot_aggregate', '/var/lib/docker/tmp', '/etc/keylime/start-agent.sh', '/usr/bin/xmlwf']
2021-10-21 12:27:40.065 - keylime.ima - WARNING - No boot_aggregate value found in allowlist, adding an empty one
2021-10-21 12:27:40.167 - keylime.tenant - DEBUG - b64_v: DDxMmgxNkDI8lHcHQy9dIEK0hd82ECF+1fxQ1HSBYtM=
/usr/lib/python3.9/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
2021-10-21 12:27:41.856 - keylime.tenant - DEBUG - Agent_quote received quote: reJwBhQB6//9UQ0eAGAAiAAvPI/LqE2r05ZYZ5zKiZo5OKQF9Aqhno0lgPpL9cFBwdgAUT3M3OVpXTGVyM0Joam05Q01uYVoAAAAABFQOOgAAAAMAAAAAASAZECMAFjY2AAAAAQALAwAAAQAge7/WJFAo1Gczl+Ez7PYMOSYOqKdL7KQxJJE7JyOfkQ1BcChe:eJwBBgH5/gAUAAsBABNaqQaV+Dpf972qoF2wuFtgH9yBHqCCu3JgRZa4Fj3qg6tXY9wnEiF0vYOrGwH/S9SS2wexorQpxzyj4iehIMn8Q+zy5UvG80BK8T4njCoG2ndNI0KxDkiklcAyDtQZ94vFPNjwrj0eTSg2mtMv0aNCb1oiWqNU0fKoLpdwIdphHXzkKUvN1VaYEM7lccSFstIxPXKJLuGKBgDT3nKuU0Yi18D5NpoxyobgJ98zrBOMh5VmCHaf1n9D/RBIYkZoullBYZC/2jbI9SXoZIyaOocMZCD0dfApVqxIv+i64d3cH9VOXBy3E0Cfc/E7gRRxJaOPnfbCXo4kvPZupof7ynqyRYCw:eJztw6ENgDAUQMH3g6zFgWADME0Iw7BAPSPgmQE8AoXoDPgmhAVQSFRNd6hoLzkBFAUI0UjYsGxPV4+nK+3ljlv/ldh3b5XRw2T6df7IEuIB8moQeQ==
2021-10-21 12:27:41.868 - keylime.tenant - DEBUG - Agent_quote received public key: -----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxtBakSxHhTUmqIrobUU6
rvLSC4nRDd0SDEfGGyBmyR9uZ50MpSlG3Vr6T1bvfEQi976Ct4+z3ssKxIxfRtap
fzr/TGjAndqt2J871EHDYYKKPIes4x3PeZquJCFxrEGj07bnk1Z50hFw+9l2t6Z4
iTcJYmT77ZCE/QDEjZghWrCfgGqtHkc1uUW77YpQOCxbz/7fjgY+SqWTqI6QQ8dG
gkCB9B8g71L1h5Rvcxzq+qudwehPiKrtxTrXuJPihlsE930RjoKM68UivKLrGr+8
kc8OLXJu2RnUazSfdusrR1tlbdgWGjnglI4SQ2Dk1A2qUX/NtHe0fs7ra+J5KWwj
mwIDAQAB
-----END PUBLIC KEY-----

2021-10-21 12:27:41.870 - keylime.tenant - DEBUG - Agent_quote received hash algorithm: sha256
2021-10-21 12:27:41.870 - keylime.tenant - DEBUG - Agent_quote received encryption algorithm: rsa
2021-10-21 12:27:41.871 - keylime.tenant - DEBUG - Agent_quote received signing algorithm: rsassa
/usr/lib/python3.9/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
2021-10-21 12:27:41.955 - keylime.registrar_client - CRITICAL - Error: unexpected http response code from Registrar Server: 404
2021-10-21 12:27:41.955 - keylime.registrar_client - CRITICAL - Response code 404: agent_id not found
2021-10-21 12:27:41.956 - keylime.tenant - WARNING - AIK not found in registrar, quote not validated
/usr/lib/python3.9/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
2021-10-21 12:27:42.536 - keylime.tenant - INFO - Agent D432FBB3-D2F1-4A97-9EF7-75BD81C00000 stopped
Traceback (most recent call last):
  File "/usr/local/bin/keylime_tenant", line 33, in <module>
    sys.exit(load_entry_point('keylime==0.0.0', 'console_scripts', 'keylime_tenant')())
  File "/usr/local/lib/python3.9/site-packages/keylime/cmd/tenant.py", line 15, in main
    tenant.main()
  File "/usr/local/lib/python3.9/site-packages/keylime/tenant.py", line 1243, in main
    mytenant.do_quote()
  File "/usr/local/lib/python3.9/site-packages/keylime/tenant.py", line 944, in do_quote
    raise e
  File "/usr/local/lib/python3.9/site-packages/keylime/tenant.py", line 897, in do_quote
    raise UserError(
keylime.tenant.UserError: TPM Quote from cloud agent is invalid for nonce: Os79ZWLer3Bhjm9CMnaZ
```

File: src/quotes_handler.rs (hunk 2 above)
Comment (stefanberger) [reply src/quotes_handler.rs]: I tried this now but I end up with this error message (patch pushed): 
```
$ cargo build
   Compiling keylime_agent v0.1.0 (/home/stefanb/dev/rust-keylime)
error[E0596]: cannot borrow data in an `Arc` as mutable
   --> src/quotes_handler.rs:271:63
    |
271 |             read_measurement_list(&ima_ml_pathbuf, nth_entry, &mut data.ima_measurement_list)?;
    |                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Arc<QuoteData>`

For more information about this error, try `rustc --explain E0596`.
error: could not compile `keylime_agent` due to previous error
```

I am not sure how do resolve this...

**Comment by lkatalin** [comment]
@stefanberger There have now been a lot of code changes and improvements, would you want to try a rebase and see if testing goes any better?

**Comment by ansasaki** [comment]
@stefanberger Hello, do you plan to continue working on this?

**Comment by stefanberger** [comment]
If someone wants to take it from here faster than I can ... please go ahead.

When I last had tested I think it ended up working just fine.

**Pull request closed without merging**